tools/dds/dds-server/dds-server.cpp

```diff
@@ -23,134 +23,202 @@ dds_server::dds_server()
+{
+    // Remove disconnected devices from devices list
+    for( auto dev_info : _devices_writers )
+    {
+        auto & dev = dev_info.second.device;
+        auto device_name = dev.get_info( RS2_CAMERA_INFO_NAME );
+
+        if( info.was_removed( dev ) )
+        {
+            devices_to_remove.push_back( device_name );
+        }
+    }
+
+    // Add new connected devices from devices list
+    for( auto && dev : info.get_new_devices() )
+    {
+        auto device_name = dev.get_info( RS2_CAMERA_INFO_NAME );
+        devices_to_add.push_back( { device_name, dev } );
+    }
+
+    bool device_change_detected = !devices_to_remove.empty() || !devices_to_add.empty();
+    return device_change_detected;
+}
+
+void dds_server::post_device_changes(
+    std::vector< std::string > devices_to_remove,
+    std::vector< std::pair< std::string, rs2::device > > devices_to_add )
+{
+    try
+    {
+        for( auto dev_to_remove : devices_to_remove )
+        {
+            remove_dds_device( dev_to_remove );
+        }
+
+        for( auto dev_to_add : devices_to_add )
+        {
+            add_dds_device( dev_to_add );
+        }
+    }
+
+    catch( ... )
+    {
+        std::cout << "Unknown error when trying to remove/add a DDS device" << std::endl;
+    }
+}
+
+void dds_server::remove_dds_device( std::string device_name )
+{
+    // deleting a device also notify the clients internally
+    _publisher->delete_datawriter( _devices_writers[device_name].data_writer );
+    std::cout << "Device '" << device_name << "' - removed" << std::endl;
+    _devices_writers.erase( device_name );
+}
+
+void dds_server::add_dds_device( std::pair< std::string, rs2::device > device_pair )
+{
+    auto dev_name = device_pair.first;
+    auto rs2_dev = device_pair.second;
+
+    if( !create_device_writer( dev_name, rs2_dev) )
+    {
+        std::cout << "Error creating a DDS writer" << std::endl;
+        return;
+    }
+
+    // Publish the device info, but only after a matching reader is found.
+    devices msg;
+    strcpy( msg.name().data(), dev_name.c_str() );
+    std::cout << "\nDevice '" << dev_name << "' - detected" << std::endl;
+    std::cout << "Looking for at least 1 matching reader... ";  // Status value will be appended to this line
+    
+    if( is_client_exist( dev_name ) )
+    {
+        std::cout << "found" << std::endl;
+        // Post a DDS message with the new added device
+        if( _devices_writers[dev_name].data_writer->write( &msg ) )
+        {
+            std::cout << "DDS device message sent!" << std::endl;
+        }
+        else
+        {
+            std::cout << "Error writing new device message for: " << dev_name << std::endl;
+        }
+    }
+    else
+    {
+        std::cout << "not found" << std::endl;
+        std::cout << "Timeout finding a reader for devices topic" << std::endl;
+    }
+}
+
+bool dds_server::is_client_exist( const std::string &device_name ) const
+{
+    int retry_cnt = 4;
```

Comment:
Why? Can you explain? Not clear that this function waits...

Reply:
Added explanation and a timeout parameter